CMakeLists.txt

```diff
@@ -0,0 +1,342 @@
+            off_t sent = 0;
+            off_t offset = 0;
+            ssize_t result = sendfile(2, 1, offset, (size_t)sent, NULL, &sent, 0);
+        }
+    "            HAVE_SENDFILE_BSD)
+
+        if(HAVE_SENDFILE_BSD)
+            set(FLAGS "${FLAGS} USE_SENDFILE_BSD HAVE_SENDFILE")
+            set(HAVE_SENDFILE TRUE)
+        endif()
+    endif()
+endif()
+
+# Detect IOCP support for Windows
+if(WIN32)
+    if(NOT HAVE_IOCP)
+        check_c_source_compiles("
+        #include <winsock2.h>
+        #include <windows.h>
+        int main(void) {
+            HANDLE iocp = CreateIoCompletionPort(INVALID_HANDLE_VALUE, NULL, 0, 0);
+            if (iocp == NULL) {
+                return 1;
+            }
+            CloseHandle(iocp);
+            return 0;
+        }
+    "            HAVE_IOCP)
+
+        if(HAVE_IOCP)
+            set(FLAGS "${FLAGS} FIO_ENGINE_IOCP HAVE_IOCP")
+        else()
+            message(FATAL_ERROR "No supported polling engine detected. Unable to compile facil.io.")
+        endif()
+    endif()
+else()
+    # Detect polling mechanisms (kqueue / epoll / poll)
+    if(NOT HAVE_KQUEUE AND NOT HAVE_EPOLL AND NOT HAVE_POLL)
+        check_c_source_compiles("
+        #define _GNU_SOURCE
+        #include <stdlib.h>
+        #include <sys/event.h>
+        int main(void) {
+            int fd = kqueue();
+        }
+    "            HAVE_KQUEUE)
+
+        if(HAVE_KQUEUE)
+            set(FLAGS "${FLAGS} FIO_ENGINE_KQUEUE HAVE_KQUEUE")
+        else()
+            check_c_source_compiles("
+            #define _GNU_SOURCE
+            #include <stdlib.h>
+            #include <stdio.h>
+            #include <sys/types.h>
+            #include <sys/stat.h>
+            #include <fcntl.h>
+            #include <sys/epoll.h>
+            int main(void) {
+                int fd = epoll_create1(EPOLL_CLOEXEC);
+            }
+        "                HAVE_EPOLL)
+
+            if(HAVE_EPOLL)
+                set(FLAGS "${FLAGS} FIO_ENGINE_EPOLL HAVE_EPOLL")
+            else()
+                check_c_source_compiles("
+                #define _GNU_SOURCE
+                #include <stdlib.h>
+                #include <poll.h>
+                int main(void) {
+                    struct pollfd plist[18];
+                    memset(plist, 0, sizeof(plist[0]) * 18);
+                    poll(plist, 1, 1);
+                }
+            "                    HAVE_POLL)
+
+                if(HAVE_POLL)
+                    set(FLAGS "${FLAGS} FIO_ENGINE_POLL HAVE_POLL")
+                else()
+                    message(FATAL_ERROR "No supported polling engine detected. Unable to compile facil.io.")
+                endif()
+            endif()
+        endif()
+    endif()
+endif()
+
+
+# Define the source files for the library
+file(GLOB_RECURSE LIB_SOURCES ${LIB_ROOT}/*.c ${LIB_ROOT}/*.cpp ${LIB_ROOT}/*.cxx ${LIB_ROOT}/*.c++)
+set(SOURCES ${LIB_SOURCES})
+
+# Optimization level
+set(OPTIMIZATION "-O3")
```

Comment:
CMake will set this (and some other basic options) based on the `CMAKE_BUILD_TYPE`). For example, a `Debug` build will set `-O0` and `-g`.

Reply:
This solve:
https://github.com/facil-io/cstl/blob/b219e44d4bb20ca376a092ea5076363ea54b7e04/CMakeLists.txt#L239-L242

Reply:
I can't tell if that's a question or a statement? 

First, as a general point: I'm not sure it makes sense to add config that CMake will add for you automatically.

Second, what you're doing here isn't the *same* as what CMake will do for you. CMake supports several build types out of the box — e.g., not just `Debug` and `Release` but also `MinSizeRel` and `RelWithDebInfo`. This approach would end up adding `-O3` to a `MinSizeRel` build — because it's "not Debug" — and that isn't really what you want.

Remember that part of the benefit of CMake is the ability to plug projects together. For example, someone could `FetchContent()` facil, and `target_link_libraries()` the code. But they might do that in a project that had a `CMAKE_BUILD_TYPE` of `MinSizeRel` or `RelWithDebInfo`. The code would still compile, etc. But the *facil* part of the code would build differently than the rest of the project, and that would be, you know, surprising.

Furthermore, it will be possible to set up facil to optionally fetch things like OpenSSL if they can't be found. In that case facil would become the parent project, and OpenSSL's CMake file would have been written with certain expectations in mind — e.g., about how build types are used.